adapters/connectad/connectad.go

```diff
@@ -0,0 +1,209 @@
+
+	if httpRes.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if httpRes.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Invalid Status Returned: %d. Run with request.debug = 1 for more info", httpRes.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(httpRes.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unable to unpackage bid response. Error: %s", err.Error()),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: "banner",
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func preprocess(request *openrtb.BidRequest) []error {
+	impsCount := len(request.Imp)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	secure := int8(0)
+
+	if request.Site != nil && request.Site.Page != "" {
+		pageURL, err := url.Parse(request.Site.Page)
+		if err == nil && pageURL.Scheme == "https" {
+			secure = int8(1)
+		}
+	}
+
+	for _, imp := range request.Imp {
+		cadExt, err := unpackImpExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		addImpInfo(&imp, &secure, cadExt)
+
+		if err := buildImpBanner(&imp); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+	}
+
+	request.Imp = resImps
+
+	return errors
+}
+
+func addImpInfo(imp *openrtb.Imp, secure *int8, cadExt *openrtb_ext.ExtImpConnectAd) {
+	imp.TagID = strconv.Itoa(cadExt.SiteID)
+	imp.Secure = secure
+
+	if cadExt.Bidfloor != 0 {
+		imp.BidFloor = cadExt.Bidfloor
+		imp.BidFloorCur = "USD"
+	}
+
+	return
+}
+
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpConnectAd, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var cadExt openrtb_ext.ExtImpConnectAd
+	if err := json.Unmarshal(bidderExt.Bidder, &cadExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Impression id=%s, has invalid Ext", imp.ID),
+		}
+	}
+
+	if cadExt.SiteID == 0 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Impression id=%s, has no siteId present", imp.ID),
+		}
+	}
+
+	return &cadExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("We need a Banner Object in the request"),
+		}
+	}
+
+	bannerCopy := *imp.Banner
```

Comment:
Consider moving the creation of the copy within the `banner.W == nil && banner.H == nil` block to avoid the memory allocation if it's not needed.